src/test/java/org/jenkinsci/plugins/workflow/steps/CoreStepTest.java

```diff
@@ -204,4 +204,88 @@ public void builderWithEnvironment() throws Exception {
         r.buildAndAssertSuccess(p);
     }
 
+    public static class BuilderWithWorkspaceRequirement extends Builder implements SimpleBuildStep {
+        @DataBoundConstructor
+        public BuilderWithWorkspaceRequirement() {
+        }
+        // TODO: Remove once the minimum core version for this plugin is 2.25x or newer.
+        public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+            fail("This method should not get called.");
+        }
+        // TODO: Mark as @Override once the minimum core version for this plugin is 2.25x or newer.
+        public void perform(@Nonnull Run<?, ?> run, @Nonnull EnvVars env, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+            listener.getLogger().println("workspace context required, but not provided!");
+        }
+        // TODO: Mark as @Override once the minimum core version for this plugin is 2.25x or newer.
+        public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull EnvVars env, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
+            listener.getLogger().println("workspace context required and provided.");
+        }
+        // While the @TextExtension supposedly limits this descriptor to the named test method, it still gets picked up
+        // via the @Symbol from anywhere. So that needs to be unique across tests.
+        @Symbol("builderWithWorkspaceRequirement")
+        @TestExtension("builderWithWorkspaceRequirement")
+        public static class DescriptorImpl extends BuildStepDescriptor<Builder> {
+            @Override
+            public boolean isApplicable(Class<? extends AbstractProject> jobType) {
+                return true;
+            }
+        }
+    }
+
+    @Issue("JENKINS-46175")
+    @Test
+    public void builderWithWorkspaceRequirement() throws Exception {
```

Comment:
All these new tests should, say,

```java
try {
    SimpleBuildStep.class.getMethod("requiresWorkspace");
} catch (NoSuchMethodException x) {
    assumeNoException("skipping on old Jenkins", x);
}
```

Reply:
Well, for the environment PR, ISTR one of the goals stated by @dwnusbaum was to allow implementers to do like these tests do, and essentially implement the new API without needing the new Jenkins.
So foo-plugin could set things up not to require a workspace, enabling use outside of a node context with a recent workflow-basic-steps, even on old Jenkins.
That's supported perfectly well by the implementation as it stands (as evidenced by the unit tests, which work fine on old Jenkins).

Reply:
> one of the goals stated by @dwnusbaum…

Oh? I must have missed that. We occasionally do this, though it is a bit messy in this case because you are still obliged to implement the old overloads.

Reply:
I'd have to go back and see if I remember it correctly.
I mean, I don't mind removing this as a feature, but it does not seem to cost much on this end, so I figure it makes sense to keep it. Unless the current approach actually breaks in any scenarios with New Jenkins - but I don't think it does.

Reply:
OK, no strong opinion either way, but I would recommend adding comments to the points where you call `xxx.getClass().getMethod(…)` to explain that

* in old Jenkins, this will either return a method defined in some concrete class, or throw NSME
* in new Jenkins, this will return the interface method (or an override, but the behavior is the same)

Reply:
I've added more comments, but only described it in general terms. Hope that is enough.